File: tests/containment.py
"""Containment harness for engine-touching e2e steps (Python half).

Contract mirrors internal/e2e (Go): every engine child runs in its own
process group (start_new_session) under a hard budget; a breach costs
the whole group (SIGKILL to -pgid, so grandchildren die too) plus an
artifact dump — the child's stderr and any step-specific dumpers —
under logs/e2e/<name>-<stamp>/. A wedge fails loudly with evidence
instead of hanging the suite.
"""

from __future__ import annotations

import contextlib
import os
import signal
import subprocess
import time
from collections.abc import Callable
from pathlib import Path

DEFAULT_ARTIFACT_ROOT = Path("logs/e2e")
TAIL_LINES = 40


class ContainedFailure(Exception):
    """A contained step failed (non-zero exit, breach, or cancel)."""

    def __init__(self, message: str, artifact_dir: Path, timed_out: bool = False):
        super().__init__(message)
        self.artifact_dir = artifact_dir
        self.timed_out = timed_out


def _tail(path: Path) -> str:
    if not path.exists():
        return ""
    lines = path.read_text(errors="replace").splitlines()
    return "\n".join(lines[-TAIL_LINES:])


def _dump(dumpers, artifact_dir: Path) -> None:
    """Run step-specific dumpers, absorbing their errors — a failing
    dumper must not mask the original failure."""
    for i, dumper in enumerate(dumpers):
        try:
            dumper(artifact_dir)
        except Exception as exc:  # noqa: BLE001 — evidence over purity
            (artifact_dir / f"dumper-{i}-error.txt").write_text(repr(exc), encoding="utf-8")


def run_step(
    name: str,
    argv: list[str],
    *,
    cwd: Path | None = None,
    budget: float = 240.0,
    artifact_root: Path | None = None,
    dumpers: tuple[Callable[[Path], None], ...] = (),
    env: dict[str, str] | None = None,
    on_start: Callable[[int], None] | None = None,
) -> Path:
    """Run one contained engine command; return the artifact dir.

    Raises ContainedFailure on budget breach or non-zero exit.
    """
    root = artifact_root or DEFAULT_ARTIFACT_ROOT
    stamp = time.strftime("%Y%m%d-%H%M%S", time.gmtime())
    artifact_dir = root / f"{name}-{stamp}"
    artifact_dir.mkdir(parents=True, exist_ok=True)
    stderr_path = artifact_dir / "stderr.log"

    with open(stderr_path, "wb") as stderr:
        proc = subprocess.Popen(
            argv,
            cwd=cwd,
            env=env,
            stdout=stderr,
            stderr=subprocess.STDOUT,
            start_new_session=True,  # the child IS the process group
        )
    if on_start is not None:
        on_start(os.getpgid(proc.pid))
    try:
        proc.wait(timeout=budget)
    except subprocess.TimeoutExpired:
        with contextlib.suppress(ProcessLookupError):
            os.killpg(os.getpgid(proc.pid), signal.SIGKILL)
        proc.wait()
        _dump(dumpers, artifact_dir)
        raise ContainedFailure(
            f"{name}: breached its {budget}s budget (artifacts: {artifact_dir})\n"
            f"{_tail(stderr_path)}",
            artifact_dir,
            timed_out=True,
        ) from None
    if proc.returncode != 0:
        _dump(dumpers, artifact_dir)
        raise ContainedFailure(
            f"{name}: exited {proc.returncode} (artifacts: {artifact_dir})\n{_tail(stderr_path)}",
            artifact_dir,
        )
    return artifact_dir


def compose_dumpers(engine: str, project: str) -> tuple[Callable[[Path], None], ...]:
    """Standard dumpers for compose steps: ps state + per-container logs."""

    def ps(artifact_dir: Path) -> None:
        _sink(engine, project, ["ps", "-a"], artifact_dir / "compose-ps.txt")

    def logs(artifact_dir: Path) -> None:
        _sink(engine, project, ["logs", "--tail", "100"], artifact_dir / "compose-logs.txt")

    return (ps, logs)


def _sink(engine: str, project: str, verb: list[str], dest: Path) -> None:
    argv = [engine, "compose", "-p", project, *verb]
    try:
        proc = subprocess.run(argv, capture_output=True, text=True, timeout=60, check=False)
        dest.write_text(proc.stdout + proc.stderr, encoding="utf-8")
    except Exception as exc:  # noqa: BLE001 — evidence over purity
        dest.write_text(repr(exc), encoding="utf-8")

File: internal/e2e/containment.go
// Package e2e is the containment harness for engine-touching tests:
// every engine child runs in its own process group under a hard
// budget, and a breach costs the whole group (SIGKILL) plus an
// artifact dump — compose state, engine logs, the child's own stderr.
// A wedge therefore fails loudly with evidence instead of hanging the
// suite. Imports nothing internal (foundation package).
package e2e

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"time"
)

// DefaultArtifactRoot is where dumps land when Step.ArtifactRoot is
// empty — the repo's existing e2e log convention.
const DefaultArtifactRoot = "logs/e2e"

// tailLines is how much of the child's output an error carries; the
// full stream always lands in the artifact dir.
const tailLines = 40

// Step is one contained engine command.
type Step struct {
	// Name prefixes the artifact directory (e.g. "plane-up").
	Name string
	// Dir is the child's working directory.
	Dir string
	// Env is the full environment (nil inherits).
	Env []string
	// Argv is the command to run.
	Argv []string
	// Budget bounds the child's runtime; breach = group SIGKILL.
	// Zero means 240s.
	Budget time.Duration
	// ArtifactRoot overrides DefaultArtifactRoot (tests use t.TempDir).
	ArtifactRoot string
	// Dumpers run inside the artifact dir after a failure — compose
	// ps, engine logs, anything that explains the corpse.
	Dumpers []func(dir string) error
	// OnStart exposes the child's process group id (tests assert the
	// group actually died).
	OnStart func(pgid int)
}

// Result reports how a step ended.
type Result struct {
	ArtifactDir string
	TimedOut    bool
}

// Run executes the step under containment. A budget breach or
// non-zero exit returns an error carrying the artifact dir and the
// output tail; ctx cancellation kills the group the same way.
func Run(ctx context.Context, s Step) error {
	if len(s.Argv) == 0 {
		return fmt.Errorf("containment: empty argv")
	}
	if s.Budget <= 0 {
		s.Budget = 240 * time.Second
	}
	root := s.ArtifactRoot
	if root == "" {
		root = DefaultArtifactRoot
	}
	stamp := time.Now().UTC().Format("20060102-150405")
	dir := filepath.Join(root, fmt.Sprintf("%s-%s", s.Name, stamp))
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("containment: artifact dir: %w", err)
	}

	cmd := exec.Command(s.Argv[0], s.Argv[1:]...)
	cmd.Dir = s.Dir
	cmd.Env = s.Env
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	outFile, err := os.Create(filepath.Join(dir, "stderr.log"))
	if err != nil {
		return fmt.Errorf("containment: stderr capture: %w", err)
	}
	defer outFile.Close()

	var mu sync.Mutex
	var lines []string
	cmd.Stdout = writerFunc(func(p []byte) (int, error) {
		mu.Lock()
		all := strings.Split(strings.TrimRight(string(p), "\n"), "\n")
		lines = append(lines, all...)
		if len(lines) > tailLines {
			lines = lines[len(lines)-tailLines:]
		}
		mu.Unlock()
		return outFile.Write(p)
	})
	cmd.Stderr = cmd.Stdout

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("containment: start: %w (artifacts: %s)", err, dir)
	}
	pgid := cmd.Process.Pid // Setpgid makes the group id == child pid
	if s.OnStart != nil {
		s.OnStart(pgid)
	}

	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()

	breach := false
	timer := time.AfterFunc(s.Budget, func() {
		breach = true
		killGroup(pgid)
	})
	select {
	case err := <-done:
		timer.Stop()
		// A racing timer may have fired just as the child exited.
		_ = breach
		if err == nil {
			return nil
		}
		s.dump(dir)
		return fmt.Errorf("containment: %s exited: %v (artifacts: %s)\n%s",
			s.Argv[0], err, dir, strings.Join(tail(&mu, lines), "\n"))
	case <-ctx.Done():
		timer.Stop()
		killGroup(pgid)
		<-done
		s.dump(dir)
		return fmt.Errorf("containment: %s cancelled (artifacts: %s)\n%s",
			s.Argv[0], dir, strings.Join(tail(&mu, lines), "\n"))
	case <-time.After(s.Budget + 5*time.Second):
		// The kill timer fired but Wait hasn't returned — a wedged
		// grandchild holding the pipe. Kill again and stop waiting.
		killGroup(pgid)
		s.dump(dir)
		return fmt.Errorf("containment: %s breached its %s budget and ignored SIGKILL (artifacts: %s)\n%s",
			s.Argv[0], s.Budget, dir, strings.Join(tail(&mu, lines), "\n"))
	}
}

// killGroup SIGKILLs every process in the group — children and
// grandchildren die with the leader, so orphans cannot outlive the
// suite.
func killGroup(pgid int) {
	_ = syscall.Kill(-pgid, syscall.SIGKILL)
}

// dump runs the step's dumpers, absorbing their errors — a failing
// dumper must not mask the original failure.
func (s Step) dump(dir string) {
	for i, d := range s.Dumpers {
		func() {
			defer func() {
				if r := recover(); r != nil {
					_ = os.WriteFile(filepath.Join(dir, fmt.Sprintf("dumper-%d-panic.txt", i)),
						[]byte(fmt.Sprint(r)), 0o644)
				}
			}()
			_ = d(dir)
		}()
	}
}

func tail(mu *sync.Mutex, lines []string) []string {
	mu.Lock()
	defer mu.Unlock()
	if len(lines) > tailLines {
		return lines[len(lines)-tailLines:]
	}
	return lines
}

type writerFunc func([]byte) (int, error)

func (f writerFunc) Write(p []byte) (int, error) { return f(p) }
